### DISCUSSION TITLE
FAQ

### DISCUSSION THREAD

**Discussion by antonionardella** [iota-community/X-Team_IOTA_SmartContracts#3] [Q&A]
Please use this discussion thread to post and discuss frequently asked questions about ISCP.

**Comment by Ticket2Andromeda** [comment]
Q: What languages will be supported for ISCP?
A: At the moment Rust is the supported language by IF for creating smart contracts. For debugging reasons, the Go programming language is used for running unit tests on smart contracts. The unofficial answer, however, is that all programming languages that are fully supported by WASM will be able to be used by the ISCP.
(You guys already know you're going to get this one a billion times)

**Reply by iamadamsiwy** [reply]
WebAssembly _AKA_ WASM will be the bytecode running on the VM.  [Further Reading](https://webassembly.org/getting-started/developers-guide/) - For a Quickstart: [Compiling Smart Contract from Rust ](https://github.com/iotaledger/wasp/blob/master/articles/tutorial/03.md#writing-and-compiling-first-rust-smart-contract)

**Comment by lukastanisic99** [comment]
Q: Why are IOTA Smart Contracts Layer 2?
A: IOTA uses a DAG on Layer 1 which is scalable by design, but doesn't have total order. Total order is essential for Smart Contracts so it's achieved on Layer 2 by taking inputs from Layer 1 and selecting a committee that uses classical BFT consensus to agree on the order of those inputs.

**Comment by lukastanisic99** [comment]
Q: Pros of IOTA Smart Contracts?
A: IOTA Smart Contracts are very flexible and can be used for various different use-cases. Multiple Committees can run in parallel without knowing about other committees and thus IOTA Smart Contracts are scalable. Committee size is flexible and depends on the use-case. The bigger the committee the higher the security. Bigger committees most probably have higher operational cost - fees for executing the smart contract. There could even exist permissioned committees. The fee structure is also flexible. Fees can be 0, but most likely they won't be in the open market of permissionless and reputable nodes that provide a stake. The stake is also not mandatory and can be 0.

**Comment by lukastanisic99** [comment]
Q: What are the tradeoffs of using IOTA Smart Contracts in comparison to Ethereum Smart Contracts?
A: Ethereum Smart Contracts are Layer 1 and thus get validated by the entire network and are considered the most secure smart contracts. Because they get validated by the entire network there scale is limited. Each Smart Contract on Ethereum has the same security and that's the security of the entire network. In IOTA this is not the case. The security of each Smart Contract is directly proportional to the security of the committee that governs it. Depending on your use-case you won't need the entire network to validate your Smart Contract. That's why IOTA Smart Contracts are flexible and scalable. They give you the option to chose the security level and the fees you're going to pay. Higher security would most probably lead to higher fees in a permissionless setting.

**Comment by lukastanisic99** [comment]
Q: Why aren't Smart Contracts free?
A: Smart Contracts can be free, but in a permissionless setting they most probably won't be. Why? Nodes have no incentive to run your Smart Contract for free. On top of that, the security is guaranteed by the funds that the committee members stake. They risk losing their funds if they don't follow protocol. Why would any participant in any system take on risk for free? They do it, because they expect a return. The higher the stake, the higher the security and thus most probably higher the fees.

**Comment by lukastanisic99** [comment]
Q: How secure are IOTA Smart Contracts?
A: They're totally secure as long as the committee of nodes are managing less funds than they have staked. All Smart Contracts are deterministic and any deviation from the protocol is easily detected. If the committee doesn't follow protocol they have collateral that can be slashed. This is true for a permissionless setting. In a permissioned setting you don't need stake because you control the nodes that govern the Smart Contract.

**Comment by iamadamsiwy** [comment]
Q: How do Wasp Nodes connect to Goshimmer Nodes?
A: IOTA Smart Contracts runs on Layer 2 deployed on multiple **Wasp Nodes**. These nodes are 'connected' to Layer 1 Goshimmer Nodes via **WaspConn Plugin**, that needs to be enabled on the Goshimmer Node.

**Comment by iamadamsiwy** [comment]
Q: What is the difference between wwallet and wasp-cli?
A: WWallet was the former name of the WASP-CLI. [Read more about wasp-cli](https://github.com/iotaledger/wasp/blob/master/tools/wasp-cli/README.md)

**Comment by iamadamsiwy** [comment]
Q: wasp-cli request-funds - Faucet request seems to have failed. What should I do?
A: _Is your goshimmer node synced? TBD_ / _Enable the Faucet Plugin in your GoShimmer Node??? TBD_

**Comment by kristofvdj88** [comment]
Q: How are misbehaving validator nodes inside a committee slashed? 
A: Misbehavior is easily detectible because of the deterministic nature of the states. The detection of this can be done by so called sentinels. Sentinels can be a part of the committee or it can be a third party. It's function is to fill in a complaint to the staking contract with the cryptographical prove of misbehavior. Sentinels are incentivized by being rewarded the slashed stake.

**Comment by kristofvdj88** [comment]
Q: If a certain node stakes higher then others, does it get more rights and privileges in the committee? 
A: Stakes do not give any special rights, they are needed as security bonds. This and the fact that the fees are distributed equally makes no sense to have different stakes for different validators in the committee. Basically the stakes are proportional to the desired security of the committee (the trust score).

**Comment by kristofvdj88** [comment]
Q: Can a committee censor requests? 
A: Committees can’t censor requests, because they always have consensus on inputs. Sure, quorum majority could censor anything of course, but that is a misbehavior. If they choose to do this and lose their stakes, nobody can prevent it and state transitions cannot be reverted on L1. However, it would be one-time action.

**Comment by kristofvdj88** [comment]
Q: Where will the validator node's staked funds reside? 
A: The staking contract holds all staked funds. This contract is executed by the top committee composed of high-mana/ high-reputation nodes and is closely tied to L1, the ones which have max skin in the game. The funds are owned and moved by the smart contract.

**Comment by kristofvdj88** [comment]
Q: Can a committee change validator nodes? 
A: The committee of nodes may rotate its members (add/remove) per epoch according to the core contract on the same chain. 
_Possibly there’s more info to be added here?_

**Comment by kristofvdj88** [comment]
Q: ISCP is a “multichain” environment. What does this mean? 
A: ISCP is a multichain environment where there are many chains, where on every chain there can be many smart contracts. In Ethereum for example there's no such thing as concurrent programming, i.e. you can't run two chains in parallel. The whole ecosystem of a blockhain is running in sync, like many ships welded together and floating in stormy waters. Multiple chains of ISCP is a trustless transacting platform where you can split your dApp into parallel pieces to run on multiple chains. For example, a distributed Oracle smart contract can run on one chain and the consumer of oracle's data (DeX) can run on another in parallel. Neither is a bottleneck in high demand situations.

**Comment by kristofvdj88** [comment]
Q: Why isn't mana used for staking? After all it does describe a node's reputation?
A: Mana is not purely reputation, but rather it’s a provable composite variable which includes owned assets, activity and, possibly, reputation. So, it reflects your “skin in the game” very well, however it cannot be slashed in case of misbehavior. Who on the L1 would do that? The L1 would have to have consensus on someone’s misbehavior on L2 and this is very problematic. So, mana cannot be staked in this strict sense.

**Comment by kristofvdj88** [comment]
Q: What are request tokens and what is their purpose?
A: Every time one performs a smart contract request, a request token must be minted (by the sender from its wallet) in a L1 transaction to the smart contract address. A request token is a single colored token that gets uncolored after the request has been handled by the smart contract so it knows it's no longer part of its backlog. The remaining 1 iota is either accounted as a part of fee, or (if fee == 0) is returned to the sender. In short, think of it an internal bookkeeping tool for the contract.

**Comment by Phyloiota** [comment]
Q: what happens if too many requests are happening to a smart contract? on something like ethereum gas price ensure it doesn't happen so i wonder how a smart contract would prevent itself getting congested?

A: here are several things. One chain can handle even 100 TPS (Ethereum is 11 TPS for the whole network iirc) . We can have many parallel chains and the total TPS is limited by the tangle throughput. IOA 2.0 has congestion control on L1. We have fees and we can also introduce gas if needed.  We also can introduce exotic strategies, something like requiring 1Mi for each request of collateral and returning it all back if same address doesn't send next request in at least 2 min.

**Comment by Phyloiota** [comment]
Q: Does every smart contract have its own chain? And in that case would these measures like fees etc be implemented by each contract or would it be network-wide?

A: No, each chain can host many smart contracts, just like any other blockchain. The contracts on one chain can interact just calling each other, like functions

**Comment by Phyloiota** [comment]
Q: What determines the transaction hash of a smart contract tx?

A: every SC transaction is a value transaction. The ID of value transaction in Goshimmer (aka tx hash) is a hash of the following: inputs, outputs, data payload (which in sc transaction contains sc metadata), user timestamp, mana pledge and signatures

**Comment by Phyloiota** [comment]
Q: does a smart contract have a build in way to keep state? If yes how large can this state be?

A: Yes. It's all about the state. A SC without state is probably not very useful.
State can be pretty large, but you want to take processing limitations and storage limitations into account.
Also, you need to bring the state data into the SC somehow. That means you are sending the data to the SC over the Tangle. So there's another limitation.

**Comment by Phyloiota** [comment]
Q: do IOTA smart contracts own addresses, and thus own IOTA if it is sent there?

A: a bit different. We have many chains and each  chain has an address on the Tangle. There are many smart contacts on each chain. Therefore smart contract is identified by chain address + contract ID on the chain , i.e. just one address is not enough  You can send (deposit) tokens to the smart contract and it will own it. If you send token to the plain chain address on the tangle, some fallback action will be invoked, for example all those tokens will be accrued to the chain owner or returned to the sender.

**Comment by Phyloiota** [comment]
Q: How does committee selection work? What prevents a bad actor from taking over an entire committee?

A: there's no automatic committee selection in ISCP. You form a consortium and run a chain. The chain is as secure as committee is.
The automatic committee selection in a permissionless setting (with staking etc) is in our future development plan

**Comment by Phyloiota** [comment]
Q: if the colour is the criteria to distinct between coloured coins, does it mean the max. possible amount is equal to the max possible hex color code combinations? 

A: This isn’t an RGB hex code if that is what you mean.
Currently it's a 32-byte transaction hash.  1.1579 x 10^77 possible combinations

**Comment by Phyloiota** [comment]
Q:  wich runtime have you choose to run wasm ?

A_ wasmtime

**Comment by Phyloiota** [comment]
Q:  So a smart contract owns iota on layer 1 or does the committee own the iota (in 1 address with shared key or something) and a smart contract only owns layer 2 tokens created by the committee (which are backed by the iota on layer 1)?

A:  Smart contract owns L1 asssets in its account on the chain. On L1 it looks like all smart contracts of then chain holds their tokens in one address, owner by the committee. However, chain logic maintains accounts in its state. A smart contract can’t own tokens contained in some other address on L1.

**Comment by Phyloiota** [comment]
Q:  how their smart contract subchains are secured by L1. Sure, the subchain state is validated by L1 and therefore resistant to reorgs. But that doesn't help me if the sidechain operator (or consortium) e.g. stops block processing and demands a ransom for my funds.

A:  The short answer: trust/security of the chain is equal to the trust to the committee. Trust to the committee is proportional to its “skin in the game”, how much it will lose for provable misbehavior (universal principle). In consortiums “skin in ghe game” is a real world reputation (if you are CocaCola, you wont misbehave to steal 1 USD) or legal commitment (court will punish you for misbehavior). For permissionless market validators are staked with assets, it means you will lose your asset if you misbehave. It is all game theory
